### PR TITLE
feat: directly build tree_r stores as cached trees

### PR DESCRIPTION
**Pull request by cryptonemo**
fix: upgrade failure dep and remove workaround

### CODE REVIEW

File: storage-proofs/src/merkle.rs [hunk 1]
```diff
@@ -315,8 +321,46 @@ pub fn create_lcmerkle_tree<H: Hasher, U: typenum::Unsigned>(
...
+    replica_path: &PathBuf,
+) -> Result<LCMerkleTree<H::Domain, H::Function, U>> {
+    trace!("open_lcmerkle_tree called with size {}", size);
+    trace!(
+        "is_merkle_tree_size_valid({}, arity {}) = {}",
+        size,
+        U::to_usize(),
+        is_merkle_tree_size_valid(size, U::to_usize())
+    );
+    assert!(is_merkle_tree_size_valid(size, U::to_usize()));
```
Comment (dignifiedquire) [inline storage-proofs/src/merkle.rs]: should be `ensure` to avoid panics

**Comment by dignifiedquire** [comment]
released `merkletree@0.16.4`

**Comment by cryptonemo** [comment]
> released `merkletree@0.16.4`

Thanks, I'll update this shortly.

File: storage-proofs/src/merkle.rs [hunk 2]
```diff
@@ -314,9 +323,56 @@ pub fn create_lcmerkle_tree<H: Hasher, U: typenum::Unsigned>(
...
+        LCMerkleTree::<H::Domain, H::Function, U>::try_from_iter_with_config(
+            (0..size).map(f),
+            config,
+        )?;
+
+    let store: &mut LevelCacheStore<_, std::fs::File> = lc_tree.data_mut();
+    store
+        .set_external_reader(
+            ExternalReader::new_from_path(replica_path)
+                .expect("Failed to create external reader from path"),
```
Comment (dignifiedquire) [inline storage-proofs/src/merkle.rs]: lets not panic here

File: storage-proofs/src/merkle.rs [hunk 3]
```diff
@@ -314,9 +323,56 @@ pub fn create_lcmerkle_tree<H: Hasher, U: typenum::Unsigned>(
...
+        )
+        .expect("Failed to set external reader");
```
Comment (dignifiedquire) [inline storage-proofs/src/merkle.rs]: lets not panic here

File: storage-proofs/src/porep/stacked/vanilla/proof.rs [hunk 4]
```diff
@@ -554,19 +556,10 @@ impl<'a, H: 'static + Hasher, G: 'static + Hasher> StackedDrg<'a, H, G> {
...
-        if tree_r_last_config.levels != 0 {
-            tree_r_last.compact(
-                tree_r_last_config.clone(),
-                StoreConfigDataVersion::Two as u32,
-            )?;
-        }
-
         data.drop_data();
 
+        let comm_r: H::Domain = H::Function::hash2(&tree_c.root(), &tree_r_last.root());
```
Comment (dignifiedquire) [inline storage-proofs/src/porep/stacked/vanilla/proof.rs]: please keep the comment

**Pull request merged**